### ISSUE TITLE
[Question] Running script from blueprint without attaching to actor

### ISSUE THREAD

**Issue by pwnpwn** [ncsoft/Unreal.js#110]
With limited documentation on the methods exposed in blueprint, I can not seem to figure out how to how to run a script file without attaching it to something.

If I attach it to something, the web-socket server I am running gets shut down on level change. So I guess attaching the script to the game instance would be ideal, but I can not seem to figure out how to do this either.

Is there a simple solution to my problem? Any help or nudge in the right direction would be greatly appreciated.

**Comment by nakosung** [comment]
You can override GameInstance class(Project settings - Project - Map & Mod) with your custom BP class. For detailed implementation you may refer to HelloGame.umap in Examples project.

**Comment by pwnpwn** [comment]
Yes, I have a custom GameInstance, but since it cannot contain actors I can not figure out how to launch a script from there. There is the "Run Script" node, but there does not seem to be a way to resolve the required parameter "Javascript context" in GameInstance.

**Comment by pwnpwn** [comment]
I am installing visual studio now, so that I can take a look at the samples. Now that you have the plugin in the marketplace, would it be possible to provide a "blueprint only" project with the samples?

**Comment by nakosung** [comment]
@pwnpwn Yes, JavascriptPlayground project is actually a blueprint only project. https://drive.google.com/open?id=0B8EsJUQpOkucRVJna3dEczJHN2M

You can create JavascriptIsolate and JavascriptContext without actors.

1. Create a `UJavascriptIsolate`.
2. Create a context by calling `isolate.CreateContext()`.
3. Expose some useful root objects by calling `context.Expose(...)`.
4. Run main script by calling `context.RunScript(filename)`.

https://github.com/ncsoft/Unreal.js-core/blob/master/Source/V8/Private/JavascriptComponent.cpp

**Comment by kelfire** [comment]
Wow.  I was wondering how to run a javascript without using the JavaComponent.  I'll pose a link to a image that does the above code.  http://imgur.com/a/fNf5I

**Comment by nakosung** [comment]
BTW, @kelfire You can access GWorld by calling `UJavascriptLibrary::Actor_GetWorld(Actor*)`.

**Comment by pwnpwn** [comment]
With your answers combined I have the functionality I was looking for.
There is a way to access the GWorld with an actor as @nakosung mentioned, but said world might not persist when changing levels:
"In a standalone game, generally only a single World exists except during seamless area transitions when both a destination and current world exists. In the editor many Worlds exist: The level being edited, each PIE instance, each editor tool which has an interactive rendered viewport, and many more."
(ue documentation)

Since I don't really need the world, only means of communicating with BP even after a level switch, I exposed the GameInstance like so:

http://imgur.com/o3EgBpG

I have not tested this extensively, but I think it will be sufficient for what I am trying to do.
I think I finally wrapped my head around how this is working, thank you so much for the help.

**Comment by pwnpwn** [comment]
btw @nakosung the project example you provided via gdrive was possible to launch without visual studio, but before that I could not find samples anywhere that would not require me to compile the plugin myself. That is what I meant by "BP only". If I did not miss it, I think It would be nice if that project was linked at github.

**Comment by nakosung** [comment]
@pwnpwn Have you installed Unreal.js as a market place plugin via epic games launcher? It contains binaries for android, mac, win32 and win64.

**Comment by pwnpwn** [comment]
I did, but the snippet editor and your examples on master wanted my to compile the plugin myself anyways. Maybe you have updated this now.

**Comment by nakosung** [comment]
Yes. Neither examples on master nor snippet editor is a bp only uproject.

2016년 9월 14일 (수) 오후 4:38, Anton Lind <notifications@github.com>님이 작성:

**Comment by Sunjammer** [comment]
This looks like it's touching on what I'm trying to solve: I need to be able to create a javascript context, evaluate some JS in it, and call methods on it to perform some processing. Is this something I can do from blueprint? It looks a little bit like to wire this stuff up I need to do most if not all of it in JS.

Conceptually:

Player hits button -> Button triggers blueprint script -> Blueprint calls a JS method -> Blueprint uses return value from JS method to affect world. 

Is this possible? I suppose what i'm mostly interested in from UnrealJS is the ability to solve isolated problems in JS without having the JS affect the rest of the running game, I suppose more of a pull-model  than push.

**Comment by nakosung** [comment]
It can be done with 'Delegates' or 'Overriding a function`. :)

**Comment by Sunjammer** [comment]
So basically making a blueprint with a dummy IO function, making an Actor with a Javascript component running a file that creates a blueprint extension that overrides that function?

**Comment by nakosung** [comment]
@Sunjammer Yes. You are right. But it doesn't have to be an actor to have a JavascriptIsolate/Context.

**Comment by Sunjammer** [comment]
Hm but how do I instantiate a blueprint subclass created in a JavascriptIsolate? So confused! :)

**Comment by nakosung** [comment]
@Sunjammer You can pass JS UCLASS to existing BP. (UClass object, not a UObject reference) And In your BP, you can instantiate 'UCLASS' to have an instance of JS-UCLASS.! :)
